packages/flutter_tools/lib/src/localizations/gen_l10n.dart

```diff
@@ -123,23 +123,36 @@ String _syntheticL10nPackagePath(FileSystem fileSystem) => fileSystem.path.join(
 // For example, if placeholders are used for plurals and no type was specified, then the type will
 // automatically set to 'num'. Similarly, if such placeholders are used for selects, then the type
 // will be set to 'String'. For such placeholders that are used for both, we should throw an error.
-List<String> generateMethodParameters(Message message, bool useNamedParameters) {
-  return message.placeholders.values.map((Placeholder placeholder) {
+List<String> generateMethodParameters(Message message, LocaleInfo? locale, bool useNamedParameters) {
+
+  // Check the compatibility of template placeholders and locale placeholders.
+  final Map<String, Placeholder>? localePlaceholders = message.localePlaceholders[locale];
+
+  return message.templatePlaceholders.entries.map((MapEntry<String, Placeholder> e) {
+    final Placeholder placeholder = e.value;
+    final Placeholder? localePlaceholder = localePlaceholders?[e.key];
+    if (localePlaceholder != null && placeholder.type != localePlaceholder.type) {
+      throw L10nException(
+          'The placeholder, ${placeholder.name}, has its "type" resource attribute set to '
+          'the "${localePlaceholder.type}" type in locale "$locale", but it is "${placeholder.type}" '
+          'in the template placeholders. For compatibility with template placeholders, change '
+          'the "type" attribute to "${placeholder.type}".');
```

Comment:
Nitpick: I think `placeholders` should be in its singular form here:

```suggestion
          'The placeholder, ${placeholder.name}, has its "type" resource attribute set to '
          'the "${localePlaceholder.type}" type in locale "$locale", but it is "${placeholder.type}" '
          'in the template placeholder. For compatibility with template placeholder, change '
          'the "type" attribute to "${placeholder.type}".');
```